File: getOrders.js
require("dotenv").config();
const { PrivyClient } = require("@privy-io/node");
const { ethers } = require("ethers");
const {
  createAuthenticatedRequest,
  hexToPrivateKey,
} = require("./orderly-auth");

// Use native fetch in Node.js 18+ or fallback to node-fetch
let fetch;
if (typeof globalThis !== "undefined" && globalThis.fetch) {
  fetch = globalThis.fetch;
} else {
  try {
    fetch = require("node-fetch");
  } catch (e) {
    throw new Error(
      "fetch is not available. Please install node-fetch: npm install node-fetch"
    );
  }
}

// Configuration
const ORDERLY_API_URL = "https://api.orderly.org"; // or https://testnet-api.orderly.org for testnet
const BROKER_ID = "woofi_pro";

/**
 * Generate Orderly account ID
 * Based on Orderly documentation: accountId = keccak256(abi.encode(address, keccak256(abi.encodePacked(brokerId))))
 *
 * @param {string} address - User wallet address
 * @param {string} brokerId - Broker ID
 * @returns {string} - Account ID as bytes32 hex string
 */
function getAccountId(address, brokerId) {
  const abiCoder = ethers.AbiCoder.defaultAbiCoder();
  // keccak256(abi.encodePacked(brokerId)) - using solidityPackedKeccak256 for ABI-encoded string
  const brokerIdHash = ethers.solidityPackedKeccak256(["string"], [brokerId]);
  // keccak256(abi.encode(address, brokerIdHash))
  return ethers.keccak256(
    abiCoder.encode(["address", "bytes32"], [address, brokerIdHash])
  );
}

/**
 * Get orders from Orderly account
 * Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/evm-api/restful-api/private/get-orders
 *
 * @param {Object} options - Configuration options
 * @param {string} options.walletId - The Privy wallet ID (required)
 * @param {string} options.walletAddress - The wallet address (optional, will be fetched if not provided)
 * @param {string} [options.symbol] - Trading symbol filter (optional)
 * @param {string} [options.side] - Order side filter: BUY or SELL (optional)
 * @param {string} [options.orderType] - Order type filter: LIMIT or MARKET (optional)
 * @param {string} [options.status] - Order status filter: NEW, CANCELLED, PARTIAL_FILLED, FILLED, REJECTED, INCOMPLETE, COMPLETED (optional)
 * @param {string} [options.orderTag] - Order tag filter (optional)
 * @param {number} [options.startTime] - Start time range (13-digit timestamp in milliseconds) (optional)
 * @param {number} [options.endTime] - End time range (13-digit timestamp in milliseconds) (optional)
 * @param {number} [options.page] - Page number (starts from 1) (optional, default: 1)
 * @param {number} [options.size] - Page size (max: 500) (optional, default: 25)
 * @param {string} [options.sortBy] - Sort by: CREATED_TIME_DESC, CREATED_TIME_ASC, UPDATED_TIME_DESC, UPDATED_TIME_ASC (optional)
 * @returns {Promise<Object>} - Orders result
 */
async function getOrders(options = {}) {
  try {
    // Validate environment variables
    const appId = process.env.PRIVY_APP_ID;
    const appSecret = process.env.PRIVY_APP_SECRET;

    if (!appId || !appSecret) {
      throw new Error(
        "Missing required environment variables. Please ensure the following are set in your .env file:\n" +
          "- PRIVY_APP_ID\n" +
          "- PRIVY_APP_SECRET"
      );
    }

    if (!options.walletId) {
      throw new Error("Wallet ID is required. Use --wallet-id <wallet_id>");
    }

    // Get Orderly credentials
    const orderlyKey = process.env.ORDERLY_KEY;
    if (!orderlyKey) {
      throw new Error(
        "ORDERLY_KEY environment variable is required (should be in format 'ed25519:...')"
      );
    }

    const orderlyPrivateKeyHex = process.env.ORDERLY_PRIVATE_KEY;
    if (!orderlyPrivateKeyHex) {
      throw new Error(
        "ORDERLY_PRIVATE_KEY environment variable is required (ed25519 private key in hex format)"
      );
    }

    // Convert hex string to Uint8Array
    const orderlyPrivateKey = hexToPrivateKey(orderlyPrivateKeyHex);

    // Initialize Privy client
    const privy = new PrivyClient({
      appId: appId,
      appSecret: appSecret,
    });

    // Configuration
    const brokerId = BROKER_ID;

    // Get wallet address if not provided
    let walletAddress = options.walletAddress;
    if (!walletAddress) {
      console.log("Fetching wallet details...");
      const wallet = await privy.wallets().get(options.walletId);
      walletAddress = wallet.address || wallet.addresses?.[0];
      if (!walletAddress) {
        throw new Error(
          "Could not determine wallet address from wallet object"
        );
      }
      console.log(`   Wallet Address: ${walletAddress}`);
    }

    // Derive Orderly account ID from wallet address and broker ID
    const accountId = getAccountId(walletAddress, brokerId);

    console.log("\nFetching orders from Orderly...");
    console.log(`   Wallet ID: ${options.walletId}`);
    console.log(`   Wallet Address: ${walletAddress}`);
    console.log(`   Account ID: ${accountId}`);

    // Build query parameters
    const queryParams = new URLSearchParams();
    if (options.symbol) queryParams.append("symbol", options.symbol);
    if (options.side) queryParams.append("side", options.side.toUpperCase());
    if (options.orderType) queryParams.append("order_type", options.orderType.toUpperCase());
    if (options.status) queryParams.append("status", options.status.toUpperCase());
    if (options.orderTag) queryParams.append("order_tag", options.orderTag);
    if (options.startTime) queryParams.append("start_t", options.startTime.toString());
    if (options.endTime) queryParams.append("end_t", options.endTime.toString());
    if (options.page) queryParams.append("page", options.page.toString());
    if (options.size) queryParams.append("size", options.size.toString());
    if (options.sortBy) queryParams.append("sort_by", options.sortBy);

    const queryString = queryParams.toString();
    const path = `/v1/orders${queryString ? `?${queryString}` : ""}`;

    if (queryString) {
      console.log(`   Filters: ${queryString}`);
    }

    // Create authenticated request
    const requestConfig = await createAuthenticatedRequest(
      "GET",
      path,
      null,
      accountId,
      orderlyKey,
      orderlyPrivateKey
    );

    // Make the request
    const response = await fetch(`${ORDERLY_API_URL}${path}`, requestConfig);
    const data = await response.json();

    if (!response.ok) {
      throw new Error(
        `Failed to get orders: ${JSON.stringify(data)}`
      );
    }

    if (!data.success) {
      throw new Error(
        `Orderly API returned error: ${JSON.stringify(data)}`
      );
    }

    console.log("\n✅ Orders retrieved successfully!");

    return {
      success: true,
      data: data.data,
      orders: data.data?.rows || [],
      meta: data.data?.meta || {},
      timestamp: data.timestamp,
      walletAddress: walletAddress,
      accountId: accountId,
    };
  } catch (error) {
    console.error("\n❌ Failed to get orders");
    if (error.message) {
      console.error("Error:", error.message);
    } else {
      console.error(error);
    }
    throw error;
  }
}

/**
 * Main execution function
 */
async function main() {
  try {
    // Parse command line arguments
    const args = process.argv.slice(2);
    const options = {};

    // Simple argument parsing
    for (let i = 0; i < args.length; i++) {
      if (args[i] === "--wallet-id" && args[i + 1]) {
        options.walletId = args[i + 1];
        i++;
      } else if (args[i] === "--wallet-address" && args[i + 1]) {
        options.walletAddress = args[i + 1];
        i++;
      } else if (args[i] === "--symbol" && args[i + 1]) {
        options.symbol = args[i + 1];
        i++;
      } else if (args[i] === "--side" && args[i + 1]) {
        options.side = args[i + 1];
        i++;
      } else if (args[i] === "--order-type" && args[i + 1]) {
        options.orderType = args[i + 1];
        i++;
      } else if (args[i] === "--status" && args[i + 1]) {
        options.status = args[i + 1];
        i++;
      } else if (args[i] === "--order-tag" && args[i + 1]) {
        options.orderTag = args[i + 1];
        i++;
      } else if (args[i] === "--start-time" && args[i + 1]) {
        options.startTime = parseInt(args[i + 1]);
        i++;
      } else if (args[i] === "--end-time" && args[i + 1]) {
        options.endTime = parseInt(args[i + 1]);
        i++;
      } else if (args[i] === "--page" && args[i + 1]) {
        options.page = parseInt(args[i + 1]);
        i++;
      } else if (args[i] === "--size" && args[i + 1]) {
        options.size = parseInt(args[i + 1]);
        i++;
      } else if (args[i] === "--sort-by" && args[i + 1]) {
        options.sortBy = args[i + 1];
        i++;
      } else if (args[i] === "--help" || args[i] === "-h") {
        console.log(`
Usage: node getOrders.js [options]

Required Options:
  --wallet-id <id>        Privy wallet ID to use (required)

Optional Options:
  --wallet-address <addr>  Wallet address (optional, will be fetched if not provided)
  --symbol <symbol>        Trading symbol filter (e.g., "PERP_ETH_USDC")
  --side <side>            Order side filter: BUY or SELL
  --order-type <type>       Order type filter: LIMIT or MARKET
  --status <status>         Order status filter: NEW, CANCELLED, PARTIAL_FILLED, FILLED, REJECTED, INCOMPLETE, COMPLETED
  --order-tag <tag>         Order tag filter
  --start-time <timestamp>  Start time range (13-digit timestamp in milliseconds)
  --end-time <timestamp>    End time range (13-digit timestamp in milliseconds)
  --page <page>             Page number (starts from 1, default: 1)
  --size <size>             Page size (max: 500, default: 25)
  --sort-by <sort>          Sort by: CREATED_TIME_DESC, CREATED_TIME_ASC, UPDATED_TIME_DESC, UPDATED_TIME_ASC
  --help, -h                 Show this help message

Status Values:
  - NEW: New orders
  - CANCELLED: Cancelled orders
  - PARTIAL_FILLED: Partially filled orders
  - FILLED: Fully filled orders
  - REJECTED: Rejected orders
  - INCOMPLETE: NEW + PARTIAL_FILLED (bundled status)
  - COMPLETED: CANCELLED + FILLED (bundled status)

Examples:
  # Get all orders for a Privy wallet
  node getOrders.js --wallet-id wal_xxx

  # Get orders for a specific symbol
  node getOrders.js --wallet-id wal_xxx --symbol "PERP_ETH_USDC"

  # Get only BUY orders
  node getOrders.js --wallet-id wal_xxx --side BUY

  # Get incomplete orders (NEW + PARTIAL_FILLED)
  node getOrders.js --wallet-id wal_xxx --status INCOMPLETE

  # Get orders with pagination
  node getOrders.js --wallet-id wal_xxx --page 1 --size 50

  # Get orders sorted by creation time (descending)
  node getOrders.js --wallet-id wal_xxx --sort-by CREATED_TIME_DESC

  # Get orders within a time range
  node getOrders.js --wallet-id wal_xxx --start-time 1653563963000 --end-time 1653564213000

Environment Variables:
  PRIVY_APP_ID              Privy App ID (required)
  PRIVY_APP_SECRET          Privy App Secret (required)
  ORDERLY_KEY               Orderly public key (ed25519:...) (required)
  ORDERLY_PRIVATE_KEY       Orderly private key in hex format (required)

Note: Orderly account ID is automatically derived from wallet address and broker ID.
        `);
        process.exit(0);
      }
    }

    const result = await getOrders(options);

    // Display orders in a formatted way
    console.log("\n📋 Orders:");
    console.log("=".repeat(100));

    if (result.orders.length === 0) {
      console.log("   No orders found.");
    } else {
      // Display pagination info
      if (result.meta.total !== undefined) {
        console.log(`\nTotal Orders: ${result.meta.total}`);
        console.log(`Page: ${result.meta.current_page || 1} of ${Math.ceil((result.meta.total || 0) / (result.meta.records_per_page || 25))}`);
        console.log(`Records per page: ${result.meta.records_per_page || 25}`);
      }

      result.orders.forEach((order, index) => {
        console.log(`\n${index + 1}. Order #${order.order_id}:`);
        console.log(`   Symbol: ${order.symbol}`);
        console.log(`   Side: ${order.side}`);
        console.log(`   Type: ${order.type}`);
        console.log(`   Status: ${order.status}`);
        console.log(`   Price: ${order.price}`);
        console.log(`   Quantity: ${order.quantity}`);
        if (order.amount !== null && order.amount !== undefined) {
          console.log(`   Amount: ${order.amount}`);
        }
        console.log(`   Executed Quantity: ${order.executed_quantity}`);
        console.log(`   Total Executed Quantity: ${order.total_executed_quantity}`);
        console.log(`   Visible Quantity: ${order.visible_quantity}`);
        console.log(`   Average Executed Price: ${order.average_executed_price}`);
        console.log(`   Total Fee: ${order.total_fee} ${order.fee_asset}`);
        if (order.client_order_id !== null && order.client_order_id !== undefined) {
          console.log(`   Client Order ID: ${order.client_order_id}`);
        }
        console.log(`   Realized PnL: ${order.realized_pnl}`);
        console.log(`   Created: ${new Date(order.created_time).toLocaleString()}`);
        console.log(`   Updated: ${new Date(order.updated_time).toLocaleString()}`);
      });
    }

    console.log("\n" + "=".repeat(100));
    console.log("\n📝 Summary:");
    console.log(`   Wallet Address: ${result.walletAddress}`);
    console.log(`   Account ID: ${result.accountId}`);
    console.log(`   Number of Orders: ${result.orders.length}`);
    if (result.meta.total !== undefined) {
      console.log(`   Total Orders: ${result.meta.total}`);
    }
    console.log(`   Timestamp: ${new Date(result.timestamp).toLocaleString()}`);

    process.exit(0);
  } catch (error) {
    console.error("\n❌ Failed to get orders:", error.message);
    process.exit(1);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

module.exports = { getOrders, getAccountId };

File: py/get_orders.py
"""
Get orders from Orderly account
Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/evm-api/restful-api/private/get-orders
"""
import os
import sys
import argparse
from dotenv import load_dotenv
from web3 import Web3
from orderly_auth import create_authenticated_request, hex_to_private_key
import requests
from urllib.parse import urlencode
from privy_utils import get_account_id, get_wallet_address
from orderly_constants import ORDERLY_API_URL, BROKER_ID

load_dotenv()


def get_orders(
    wallet_id: str,
    symbol: str = None,
    side: str = None,
    order_type: str = None,
    status: str = None,
    order_tag: str = None,
    start_time: int = None,
    end_time: int = None,
    page: int = 1,
    size: int = 25,
    sort_by: str = None,
) -> dict:
    """Get orders from Orderly account"""
    app_id = os.getenv("PRIVY_APP_ID")
    app_secret = os.getenv("PRIVY_APP_SECRET")
    orderly_key = os.getenv("ORDERLY_KEY")
    orderly_private_key_hex = os.getenv("ORDERLY_PRIVATE_KEY")
    
    if not app_id or not app_secret:
        raise ValueError("Missing PRIVY_APP_ID or PRIVY_APP_SECRET")
    if not wallet_id:
        raise ValueError("Wallet ID is required")
    if not orderly_key:
        raise ValueError("ORDERLY_KEY environment variable is required")
    if not orderly_private_key_hex:
        raise ValueError("ORDERLY_PRIVATE_KEY environment variable is required")
    
    orderly_private_key = hex_to_private_key(orderly_private_key_hex)
    
    print("Fetching wallet details...")
    wallet_address = get_wallet_address(wallet_id, app_id, app_secret)
    print(f"   Wallet Address: {wallet_address}")
    
    account_id = get_account_id(wallet_address, BROKER_ID)
    
    print("\nFetching orders from Orderly...")
    print(f"   Wallet ID: {wallet_id}")
    print(f"   Wallet Address: {wallet_address}")
    print(f"   Account ID: {account_id}")
    
    # Build query parameters
    query_params = {}
    if symbol:
        query_params["symbol"] = symbol
    if side:
        query_params["side"] = side.upper()
    if order_type:
        query_params["order_type"] = order_type.upper()
    if status:
        query_params["status"] = status.upper()
    if order_tag:
        query_params["order_tag"] = order_tag
    if start_time:
        query_params["start_t"] = str(start_time)
    if end_time:
        query_params["end_t"] = str(end_time)
    if page:
        query_params["page"] = str(page)
    if size:
        query_params["size"] = str(size)
    if sort_by:
        query_params["sort_by"] = sort_by
    
    query_string = urlencode(query_params)
    path = f"/v1/orders?{query_string}" if query_string else "/v1/orders"
    
    if query_string:
        print(f"   Filters: {query_string}")
    
    # Create authenticated request
    request_config = create_authenticated_request(
        "GET",
        path,
        None,
        account_id,
        orderly_key,
        orderly_private_key
    )
    
    # Make the request
    response = requests.get(
        f"{ORDERLY_API_URL}{path}",
        headers=request_config["headers"]
    )
    
    data = response.json()
    
    if not response.ok:
        raise Exception(f"Failed to get orders: {data}")
    
    if not data.get("success"):
        raise Exception(f"Orderly API returned error: {data}")
    
    print("\n✅ Orders retrieved successfully!")
    
    return {
        "success": True,
        "data": data.get("data"),
        "orders": data.get("data", {}).get("rows", []),
        "meta": data.get("data", {}).get("meta", {}),
        "timestamp": data.get("timestamp"),
        "walletAddress": wallet_address,
        "accountId": account_id,
    }


def main():
    parser = argparse.ArgumentParser(description="Get orders from Orderly account")
    parser.add_argument("--wallet-id", required=True, help="Privy wallet ID to use (required)")
    parser.add_argument("--symbol", help="Trading symbol filter (e.g., 'PERP_ETH_USDC')")
    parser.add_argument("--side", choices=["BUY", "SELL"], help="Order side filter: BUY or SELL")
    parser.add_argument("--order-type", choices=["LIMIT", "MARKET"], help="Order type filter: LIMIT or MARKET")
    parser.add_argument("--status", help="Order status filter: NEW, CANCELLED, PARTIAL_FILLED, FILLED, REJECTED, INCOMPLETE, COMPLETED")
    parser.add_argument("--order-tag", help="Order tag filter")
    parser.add_argument("--start-time", type=int, help="Start time range (13-digit timestamp in milliseconds)")
    parser.add_argument("--end-time", type=int, help="End time range (13-digit timestamp in milliseconds)")
    parser.add_argument("--page", type=int, default=1, help="Page number (starts from 1, default: 1)")
    parser.add_argument("--size", type=int, default=25, help="Page size (max: 500, default: 25)")
    parser.add_argument("--sort-by", help="Sort by: CREATED_TIME_DESC, CREATED_TIME_ASC, UPDATED_TIME_DESC, UPDATED_TIME_ASC")
    
    args = parser.parse_args()
    
    try:
        result = get_orders(
            wallet_id=args.wallet_id,
            symbol=args.symbol,
            side=args.side,
            order_type=args.order_type,
            status=args.status,
            order_tag=args.order_tag,
            start_time=args.start_time,
            end_time=args.end_time,
            page=args.page,
            size=args.size,
            sort_by=args.sort_by,
        )
        
        # Display orders
        print("\n📋 Orders:")
        print("=" * 100)
        
        if not result["orders"]:
            print("   No orders found.")
        else:
            if result["meta"].get("total") is not None:
                print(f"\nTotal Orders: {result['meta']['total']}")
                total_pages = (result["meta"].get("total", 0) + result["meta"].get("records_per_page", 25) - 1) // result["meta"].get("records_per_page", 25)
                print(f"Page: {result['meta'].get('current_page', 1)} of {total_pages}")
                print(f"Records per page: {result['meta'].get('records_per_page', 25)}")
            
            for index, order in enumerate(result["orders"], 1):
                print(f"\n{index}. Order #{order.get('order_id')}:")
                print(f"   Symbol: {order.get('symbol')}")
                print(f"   Side: {order.get('side')}")
                print(f"   Type: {order.get('type')}")
                print(f"   Status: {order.get('status')}")
                print(f"   Price: {order.get('price')}")
                print(f"   Quantity: {order.get('quantity')}")
                if order.get("amount") is not None:
                    print(f"   Amount: {order.get('amount')}")
                print(f"   Executed Quantity: {order.get('executed_quantity')}")
                print(f"   Total Executed Quantity: {order.get('total_executed_quantity')}")
                print(f"   Visible Quantity: {order.get('visible_quantity')}")
                print(f"   Average Executed Price: {order.get('average_executed_price')}")
                print(f"   Total Fee: {order.get('total_fee')} {order.get('fee_asset')}")
                if order.get("client_order_id") is not None:
                    print(f"   Client Order ID: {order.get('client_order_id')}")
                print(f"   Realized PnL: {order.get('realized_pnl')}")
                print(f"   Created: {order.get('created_time')}")
                print(f"   Updated: {order.get('updated_time')}")
        
        print("\n" + "=" * 100)
        print("\n📝 Summary:")
        print(f"   Wallet Address: {result['walletAddress']}")
        print(f"   Account ID: {result['accountId']}")
        print(f"   Number of Orders: {len(result['orders'])}")
        if result["meta"].get("total") is not None:
            print(f"   Total Orders: {result['meta']['total']}")
        print(f"   Timestamp: {result.get('timestamp', 'N/A')}")
        
        sys.exit(0)
    except Exception as error:
        print(f"\n❌ Failed to get orders: {error}")
        sys.exit(1)


if __name__ == "__main__":
    main()

